File: src/state.py
"""
state.py
--------
Session-state helpers that persist checkbox values across Streamlit reruns
for the current period only.  When the period key changes (e.g. a new day
starts) all previous keys are simply ignored and tasks reappear naturally.

On first load each day, previously saved tasks are restored from Supabase so
that goals checked in an earlier session stay hidden without the user having
to re-tick them.
"""

import datetime as dt

import streamlit as st

from src.config import ALL_TASKS
from src.date_utils import get_period_key


def _state_key(category: str, task: str, period_key: str) -> str:
    """Build a namespaced session-state key that is unique per category, task and period."""
    return f"checked|{category}|{period_key}|{task}"


def init_state(now: dt.datetime) -> None:
    """
    Ensure every task for the current period has a session-state entry.

    On the very first call of the day (detected via a sentinel key) today's
    row is fetched from Supabase and any already-saved 1s are restored into
    session state so previously checked goals stay hidden across sessions.
    Subsequent reruns within the same session skip the Supabase fetch.
    """
    from src.db import fetch_today_row, get_completed_tasks_from_row  # avoid circular import at module level

    today_sentinel = f"_loaded_{now.strftime('%Y-%m-%d')}"

    if today_sentinel not in st.session_state:
        # First load of this day in this browser session — restore from DB
        row = fetch_today_row(now)
        already_done = get_completed_tasks_from_row(row)

        for category, tasks in ALL_TASKS.items():
            pk = get_period_key(category, now)
            for task in tasks:
                k = _state_key(category, task, pk)
                if k not in st.session_state:
                    st.session_state[k] = task in already_done.get(category, [])

        # Only set the sentinel if the DB fetch actually returned data (or
        # Supabase is disabled).  If the fetch returned an empty dict AND
        # Supabase is enabled, it might be a transient error — don't lock in
        # the "already loaded" sentinel so the next rerun retries.
        from src.db import SUPABASE_ENABLED as _ENABLED
        if not _ENABLED or row:
            st.session_state[today_sentinel] = True
    else:
        # Subsequent reruns – only initialise keys that don't exist yet
        for category, tasks in ALL_TASKS.items():
            pk = get_period_key(category, now)
            for task in tasks:
                k = _state_key(category, task, pk)
                if k not in st.session_state:
                    st.session_state[k] = False


def is_checked(category: str, task: str, period_key: str) -> bool:
    """Return True if the task has been ticked in the current period."""
    return bool(st.session_state.get(_state_key(category, task, period_key), False))


def mark_checked(category: str, task: str, period_key: str) -> None:
    """Programmatically mark a task as checked (used after rerun guard)."""
    st.session_state[_state_key(category, task, period_key)] = True

File: web_dev/src/state.ts
/**
 * state.ts
 * --------
 * Global checklist state managed by Zustand.
 * Replaces Streamlit session_state — shared across all tabs without prop drilling.
 *
 * State key format: "category|periodKey|task"  (mirrors Python _state_key())
 *
 * On first load each day `initState()` fetches period rows from Supabase and
 * restores already-saved tasks so they stay hidden across app opens.
 * A date sentinel in AsyncStorage prevents redundant fetches within the same day.
 */

import AsyncStorage from "@react-native-async-storage/async-storage";
import { create } from "zustand";

import { ALL_TASKS, CATEGORIES } from "./config";
import type { Category } from "./config";
import { getPeriodKey } from "./date_utils";
import {
  fetchPeriodRows,
  getCompletedTasksFromRows,
  SUPABASE_ENABLED,
} from "./db";
import type { ChecklistState } from "./types";

// ---------------------------------------------------------------------------
// Key helpers (mirrors _state_key() in state.py)
// ---------------------------------------------------------------------------

export function stateKey(
  category: Category,
  task: string,
  periodKey: string,
): string {
  return `${category}|${periodKey}|${task}`;
}

// ---------------------------------------------------------------------------
// Zustand store
// ---------------------------------------------------------------------------

interface ChecklistStore {
  /** Flat map of stateKey → checked boolean for the current period. */
  state: ChecklistState;
  /** Whether initState has finished for today. */
  initialised: boolean;
  /** Any error from the last Supabase operation, shown as a banner. */
  lastError: string | null;

  /** Set a single task as checked. */
  markChecked: (category: Category, task: string, periodKey: string) => void;
  /** Read whether a task is checked. */
  isChecked: (category: Category, task: string, periodKey: string) => boolean;
  /** Set the last error message (cleared by passing null). */
  setLastError: (msg: string | null) => void;
  /**
   * Load state for today.  Fetches Supabase period rows on first call of the
   * day (detected via an AsyncStorage sentinel), then populates the store.
   * Subsequent calls within the same day are a no-op.
   */
  initState: (now: Date) => Promise<void>;
}

const SENTINEL_PREFIX = "@goaltracker/loaded_";

export const useChecklistStore = create<ChecklistStore>((set, get) => ({
  state: {},
  initialised: false,
  lastError: null,

  markChecked(category, task, periodKey) {
    const key = stateKey(category, task, periodKey);
    set((s) => ({ state: { ...s.state, [key]: true } }));
  },

  isChecked(category, task, periodKey) {
    return get().state[stateKey(category, task, periodKey)] ?? false;
  },

  setLastError(msg) {
    set({ lastError: msg });
  },

  async initState(now) {
    const todayStr = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}-${String(now.getDate()).padStart(2, "0")}`;
    const sentinelKey = SENTINEL_PREFIX + todayStr;

    // Build the initial flat state with all tasks set to false
    const fresh: ChecklistState = {};
    for (const category of CATEGORIES) {
      const pk = getPeriodKey(category, now);
      for (const task of ALL_TASKS[category]) {
        fresh[stateKey(category, task, pk)] = false;
      }
    }

    // Check if we already restored from DB today
    const alreadyLoaded = await AsyncStorage.getItem(sentinelKey);
    if (alreadyLoaded) {
      // Merge fresh keys that may not exist (e.g. after a task list update)
      set((s) => {
        const merged: ChecklistState = { ...fresh, ...s.state };
        return { state: merged, initialised: true };
      });
      return;
    }

    // First load of this day — restore from Supabase
    const rows = await fetchPeriodRows(now);
    const alreadyDone = getCompletedTasksFromRows(rows, now);

    for (const category of CATEGORIES) {
      const pk = getPeriodKey(category, now);
      for (const task of ALL_TASKS[category]) {
        const done = alreadyDone[category].includes(task);
        fresh[stateKey(category, task, pk)] = done;
      }
    }

    // Persist the sentinel only when DB returned data OR Supabase is disabled.
    // If the fetch returned nothing AND Supabase is enabled, a transient error
    // may have occurred — don't lock in the sentinel so the next open retries.
    if (!SUPABASE_ENABLED || rows.length > 0) {
      await AsyncStorage.setItem(sentinelKey, "1");
    }

    set((s) => {
      // Preserve any in-session checks that happened before initState resolved
      const merged: ChecklistState = { ...fresh, ...s.state };
      return { state: merged, initialised: true };
    });
  },
}));
